### PR TITLE
[GSoC'24] Enhacement: add Audio and Video multimedia options

### LINKED ISSUE #16790: [BUG] Add Image From Camera
* Add Image From Camera
* Don't Crop
* Reselect
* Press "Back"
* Save Note

=> Image is not displayed

Screenshot 2024-07-28 at 17 11 34

_Originally posted by @david-allison in https://github.com/ankidroid/Anki-Android/pull/16769#pullrequestreview-2203531034_

### PR DESCRIPTION
**Pull request by criticalAY**
## Purpose / Description
This PR enables the Audio/Video media option in multimedia selection and refactor some of the previous code
- ViewModel: using media instead of image, hence generalizing the string 
- Extracting menu provider so that it can be used in other classes of multimedia fragments

## Fixes
* Fixes #16790

## How Has This Been Tested?
Tested on Oneplus Nord ce
image
image

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by criticalAY** [comment]
Need better suggestion for the string in case no media file is selected

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+import com.ichi2.anki.multimedia.MultimediaActivity.Companion.MULTIMEDIA_RESULT_FIELD_INDEX
+import com.ichi2.anki.multimedia.MultimediaUtils.createCachedFile
+import com.ichi2.anki.utils.ext.sharedPrefs
+import com.ichi2.compat.CompatHelper
+import com.ichi2.utils.ExceptionUtil.executeSafe
+import com.ichi2.utils.FileUtil
+import timber.log.Timber
+import java.io.File
+
+class AudioVideoFragment : MultimediaFragment(R.layout.fragment_audio_video) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:52]: comment would be useful

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+        get() = if (selectedMediaOptions == MediaOption.AUDIO_CLIP) {
+            resources.getString(R.string.multimedia_editor_popup_audio_clip)
+        } else {
+            resources.getString(R.string.multimedia_editor_popup_video_clip)
+        }
+
+    private val viewModel: MultimediaViewModel by viewModels()
+
+    private val pickMediaLauncher =
+        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:63]: Does this work correctly if a user picks media, then the second pick fails?

Only some of the `View.GONE` changes are reverted

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+import com.ichi2.anki.utils.ext.sharedPrefs
+import com.ichi2.compat.CompatHelper
+import com.ichi2.utils.ExceptionUtil.executeSafe
+import com.ichi2.utils.FileUtil
+import timber.log.Timber
+import java.io.File
+
+class AudioVideoFragment : MultimediaFragment(R.layout.fragment_audio_video) {
+    override val title: String
+        get() = if (selectedMediaOptions == MediaOption.AUDIO_CLIP) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:54]: nit: I would move this to `selectedMediaOptions.toTitle(context)`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+    override fun onStop() {
+        super.onStop()
+        Timber.d("Stopping media player")
+        mediaPlayer.playWhenReady = false
+    }
+
+    companion object {
+
+        private var selectedMediaOptions: MediaOption = MediaOption.UNKNOWN
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:380]: Don't put variables in `companion object` (global variables are bad), pass them to the instance

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+                multimediaExtra
+            )
+        }
+    }
+
+    /** Enum class representing various media file options */
+    enum class MediaOption {
+        AUDIO_CLIP,
+        VIDEO_CLIP,
+        UNKNOWN
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:400]: Don't have `UNKNOWN`, make it null

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+            selectedMediaOptions = mediaOptions
+            return MultimediaActivity.getIntent(
+                context,
+                AudioVideoFragment::class,
+                multimediaExtra
+            )
+        }
+    }
+
+    /** Enum class representing various media file options */
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:396]: Don't mention `enum class`, just what it represents

Probably:

`The supported media types which [AudioVideoFragment]` supports

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt [hunk 7]
```diff
@@ -66,6 +72,28 @@ abstract class MultimediaFragment(@LayoutRes layout: Int) : Fragment(layout) {
         }
     }
 
+    /**
+     * Attaches a `MenuProvider` to the activity for creating its menu.
+     *
+     * @param menuProvider An instance of the `MenuProvider` class that will be responsible for inflating and configuring the menu.
+     */
+    fun setupMenu(menuProvider: MenuProvider) {
+        (requireActivity() as MenuHost).addMenuProvider(menuProvider)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt:81]: nit (probably for another cleanup PR): I'd see if we can move this to `requireAnkiActivity().addMenuProvider()`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt [hunk 8]
```diff
@@ -66,6 +72,28 @@ abstract class MultimediaFragment(@LayoutRes layout: Int) : Fragment(layout) {
...
+     * This method first uses `Formatter.formatFileSize(context, length)` to get the formatted file
+     * size string based on the provided `file` and the current context retrieved using `requireContext()`.
+     *
+     * @param file The `File` object representing the media file for which to get the size.
+     * @param textView The resource ID of the TextView widget where the file size will be displayed. This parameter should be annotated with `@IdRes`.
+     */
+    fun displayMediaFileSize(file: File, @IdRes textView: Int) {
+        val size = Formatter.formatFileSize(requireContext(), file.length())
+        view?.findViewById<TextView>(textView)?.text = size
+    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt:95]: This isn't a good method, it combines two separate concerns

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt:95]: Also note that `Formatter.formatFileSize` has different implementations (K = 1024 -> 1000) before Android O

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+     * @param clipCopy The File object representing the temporary file where the media clip will be copied.
+     */
+    private fun copyMediaFileToTemp(selectedMediaClip: Uri, clipCopy: File) {
+        try {
+            requireContext().contentResolver.openInputStream(selectedMediaClip).use { inputStream ->
+                CompatHelper.compat.copyFile(inputStream!!, clipCopy.absolutePath)
+
+                viewModel.currentMultimediaPath = clipCopy.absolutePath
+                viewModel.selectedMultimediaLength = File(clipCopy.absolutePath).length()
+                displayMediaFileSize(clipCopy.absoluteFile, R.id.media_size_textview)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:357]: Is `absoluteFile` necessary? If so, comment

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt (hunk 8 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt:95]: But better document the method I added 

```patch
Subject: [PATCH]
---
Index: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt b/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt	(revision 76d7ca232cd6aa3b937cad5e559a79acfe1f7d0f)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt	(date 1721580205063)
@@ -103,6 +103,7 @@
 
     private lateinit var mediaPlayer: ExoPlayer
     private lateinit var playerView: PlayerView
+    private lateinit var mediaFileSize: TextView
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
@@ -160,7 +161,7 @@
         playerView = requireView().findViewById(R.id.player_view)
         mediaPlayer = ExoPlayer.Builder(requireContext()).build()
         playerView.player = mediaPlayer
-
+        mediaFileSize = requireView().findViewById(R.id.media_size_textview)
         playerView.setControllerAnimationEnabled(true)
 
         if (selectedMediaOptions == MediaOption.AUDIO_CLIP) {
@@ -354,7 +355,7 @@
 
                 viewModel.currentMultimediaPath = clipCopy.absolutePath
                 viewModel.selectedMultimediaLength = File(clipCopy.absolutePath).length()
-                displayMediaFileSize(clipCopy.absoluteFile, R.id.media_size_textview)
+                mediaFileSize.text = clipCopy.toHumanReadableSize()
             }
         } catch (e: Exception) {
             Timber.e(e, "Unable to copy media file from ContentProvider")
Index: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt b/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt	(revision 76d7ca232cd6aa3b937cad5e559a79acfe1f7d0f)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt	(date 1721580205066)
@@ -20,8 +20,6 @@
 import android.os.Bundle
 import android.text.format.Formatter
 import android.view.View
-import android.widget.TextView
-import androidx.annotation.IdRes
 import androidx.annotation.LayoutRes
 import androidx.appcompat.app.AlertDialog
 import androidx.core.view.MenuHost
@@ -81,18 +79,7 @@
         (requireActivity() as MenuHost).addMenuProvider(menuProvider)
     }
 
-    /**
-     * Displays the human-readable file size of a media file in the specified TextView.
-     * This method first uses `Formatter.formatFileSize(context, length)` to get the formatted file
-     * size string based on the provided `file` and the current context retrieved using `requireContext()`.
-     *
-     * @param file The `File` object representing the media file for which to get the size.
-     * @param textView The resource ID of the TextView widget where the file size will be displayed. This parameter should be annotated with `@IdRes`.
-     */
-    fun displayMediaFileSize(file: File, @IdRes textView: Int) {
-        val size = Formatter.formatFileSize(requireContext(), file.length())
-        view?.findViewById<TextView>(textView)?.text = size
-    }
+    fun File.toHumanReadableSize(): String = Formatter.formatFileSize(requireContext(), this.absoluteFile.length())
 
     /**
      * Shows a Snackbar at the bottom of the screen with a predefined
Index: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt b/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt	(revision 76d7ca232cd6aa3b937cad5e559a79acfe1f7d0f)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt	(date 1721579338959)
@@ -59,6 +59,7 @@
         get() = resources.getString(R.string.multimedia_editor_popup_image)
 
     private lateinit var imagePreview: ImageView
+    private lateinit var imageFileSize: TextView
 
     private val viewModel: MultimediaViewModel by viewModels()
 
@@ -147,6 +148,7 @@
         super.onViewCreated(view, savedInstanceState)
         setupMenu(multimediaMenu)
         imagePreview = view.findViewById(R.id.image_preview)
+        imageFileSize = view.findViewById(R.id.image_size_textview)
         handleSelectedImageOptions()
         setupDoneButton()
     }
@@ -240,7 +242,7 @@
     private fun updateAndDisplayImageSize(imagePath: String) {
         val file = File(imagePath)
         viewModel.selectedMultimediaLength = file.length()
-        displayMediaFileSize(file, R.id.image_size_textview)
+        imageFileSize.text = file.toHumanReadableSize()
     }
 
     private fun showLargeFileCropDialog(length: Float) {
```

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt [hunk 10]
```diff
@@ -89,12 +83,56 @@ class MultimediaImageFragment : MultimediaFragment(R.layout.fragment_multimedia_
...
+                    when (selectedImageOptions) {
+                        ImageOptions.GALLERY -> {
+                            openGallery()
+                        }
+
+                        ImageOptions.CAMERA -> {
+                            dispatchCamera()
+                        }
+
+                        ImageOptions.UNKNOWN -> {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt:123]: Again, probably best to remove `UNKNOWN`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt [hunk 11]
```diff
@@ -133,16 +171,16 @@ class MultimediaImageFragment : MultimediaFragment(R.layout.fragment_multimedia_
     private fun setupDoneButton() {
         view?.findViewById<MaterialButton>(R.id.action_done)?.setOnClickListener {
             Timber.d("MultimediaImageFragment:: Done button pressed")
-            if (viewModel.getImageLength() == 0L) {
+            if (viewModel.getMultimediaLength() == 0L) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt:174]: What does `length` mean?

`duration` or `fileSize`

File: AnkiDroid/src/main/res/values/16-multimedia-editor.xml [hunk 12]
```diff
@@ -75,5 +75,6 @@
 
     <string name="no_image_preview">No image selected, select an image to proceed.</string>
     <string name="restart">Restart</string>
+    <string name="no_media_preview">Audio or Video not selected, select an audio or video to proceed.</string>
```
Comment (david-allison) [inline AnkiDroid/src/main/res/values/16-multimedia-editor.xml:78]: Where is this shown? It'll want to be updated before release

**Review by david-allison — CHANGES_REQUESTED** [review]
Only done a brief first-look through the code, really well designed!

**Comment by snowtimeglass** [comment]
Is it undesirable to eliminate the chances themselves that these "(media) not selected" strings and screens are shown?

(Is the elimination not feasible by showing the potentially-second bottom sheet  (app picker) in the initial screen ("Add" screen or "Edit note" screen) directly?)




Regarding the case of not selecting a media file in the picker, I guess it might be enough to just go back to the initial screen ("Add" screen or "Edit note" screen) or the initial bottom sheet, because the case is irregular.

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt (hunk 2 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:63]: I uodated the code to retain the previous media files in that case

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt (hunk 8 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt:95]: Thanks!

File: AnkiDroid/src/main/res/values/16-multimedia-editor.xml (hunk 12 above)
Comment (criticalAY) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml:78]: I am taking the suggestion from @snowtimeglass and removing the string

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt (hunk 7 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt:81]: Leaving a TODO

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt (hunk 9 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:357]: Not required replaced

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+    }
+
+    private lateinit var mediaPlayer: ExoPlayer
+    private lateinit var playerView: PlayerView
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        ankiCacheDirectory = FileUtil.getAnkiCacheDirectory(requireContext(), "temp-media")
+        if (ankiCacheDirectory == null) {
+            showSomethingWentWrong()
```
[the remark this answers is not in the document]
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt]: Resolved

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt [hunk 14]
```diff
@@ -66,6 +71,18 @@ abstract class MultimediaFragment(@LayoutRes layout: Int) : Fragment(layout) {
...
+     * Attaches a `MenuProvider` to the activity for creating its menu.
+     *
+     * @param menuProvider An instance of the `MenuProvider` class that will be responsible for inflating and configuring the menu.
+     */
+    // TODO: move this to requireAnkiActivity().addMenuProvider()
+    fun setupMenu(menuProvider: MenuProvider) {
+        (requireActivity() as MenuHost).addMenuProvider(menuProvider)
+    }
+
+    fun File.toHumanReadableSize(): String = Formatter.formatFileSize(requireContext(), this.absoluteFile.length())
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt:84]: `absoluteFile` may not be necessary, I copied it from your code

(I didn't make it explicit in my patch, my mistake)

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt (hunk 9 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt]: Should be removed in `toHumanReadableSize` as well

File: AnkiDroid/src/main/res/values/16-multimedia-editor.xml [hunk 15]
```diff
@@ -73,7 +73,6 @@
     <string name="activity_result_unexpected">App returned an unexpected value. You may need to use a different app</string>
     <string name="audio_recording_field_list" comment="Displays a list of the field data in the note editor while audio recording is taking place">Field Contents</string>
 
-    <string name="no_image_preview">No image selected, select an image to proceed.</string>
     <string name="restart">Restart</string>
```
Comment (snowtimeglass) [inline AnkiDroid/src/main/res/values/16-multimedia-editor.xml:76]: "Restart" might be somewhat confusing for users and misleading for translators. How about "Reselect".


```suggestion
    <string name="reselect">Reselect</string>
```

Comment (criticalAY) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml:76]: Reasonable change thanks 👍🏼

Comment (snowtimeglass) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml:76]: Thank you, too.

By the way, could you please guide me as to what has been done for the "Crop" icon (and its feature) in the "Add image" screen?
image
(Image from https://github.com/ankidroid/Anki-Android/pull/16673#issue-2381557104)

Comment (criticalAY) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml]: We use canhub for cropping images, so it requires the current image URI and I extracted the code for cropping images to `ImageUtils` so we use that to handle the crop result

Comment (david-allison) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml]: You should use the `replace image` material icon:

Screenshot 2024-07-23 at 11 53 42

https://fonts.google.com/icons?selected=Material+Symbols+Outlined:replace_image:FILL@0;wght@400;GRAD@0;opsz@24&icon.query=replace&icon.size=24&icon.color=%23e8eaed

Comment (snowtimeglass) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml]: @criticalAY so, after the coding, an icon for cropping is going to be revived?

Comment (snowtimeglass) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml]: I'm trying to say...
I saw "Crop" icon in a screenshot of #16673;
image

while I still don't see the icon in the same screen in a debug build from [73dde8a](https://github.com/ankidroid/Anki-Android/tree/73dde8aef587460c059e0de622e90eb6742fc1af)

image
So I'm currently supposing the "Crop" icon has temporarily been removed for some reason, and I would like to know whether the icon (and its function) in the screen can be used again in near future.

Comment (criticalAY) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml]: Ahh it's working fine in the latest commit, I have fixed that

Comment (snowtimeglass) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml]: Oh, that's nice. Thank you for letting me know.

Comment (criticalAY) [reply AnkiDroid/src/main/res/values/16-multimedia-editor.xml]: @david-allison applied the suggestion for all current media types i.e. audio, video, image

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt [hunk 16]
```diff
@@ -66,6 +74,22 @@ abstract class MultimediaFragment(@LayoutRes layout: Int) : Fragment(layout) {
...
+    fun setMenuItemIcon(menuItem: MenuItem, @DrawableRes icon: Int) {
+        menuItem.icon = ContextCompat.getDrawable(requireContext(), icon)
+    }
+
+    /**
+     * Attaches a `MenuProvider` to the activity for creating its menu.
+     *
+     * @param menuProvider An instance of the `MenuProvider` class that will be responsible for inflating and configuring the menu.
+     */
+    // TODO: move this to requireAnkiActivity().addMenuProvider()
```
[the remark this answers is not in the document]
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaFragment.kt:86]: Will create a separate PR for this

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaViewModel.kt [hunk 17]
```diff
@@ -28,27 +30,37 @@ class MultimediaViewModel : ViewModel() {
...
 
     fun setMultimediaAction(action: MultimediaBottomSheet.MultimediaAction) {
         viewModelScope.launch {
             multimediaAction.emit(action)
         }
     }
 
-    fun getImageLength(): Long {
-        return selectedImageLength
+    fun getMultimediaFileSize(): Long {
```
[the remark this answers is not in the document]
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaViewModel.kt:50]: Simplified

**Review by david-allison — CHANGES_REQUESTED** [review]
* Add Image From Camera
* Don't Crop
* Reselect
* Press "Back"
* Save Note

=> Image is not displayed

Screenshot 2024-07-28 at 17 11 34

**Comment by criticalAY** [comment]
@david-allison Fixed it, cause was if the camera was dispatched again it caused an issue due to viewModel updating the URI
` viewModel.updateCurrentMultimediaPath(it.absolutePath)`

**Review by david-allison — CHANGES_REQUESTED** [review]
* "Done" does nothing on 'Add audio clip'

API 30 emulator

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt [hunk 18]
```diff
@@ -37,16 +38,14 @@ import kotlin.reflect.KClass
 import kotlin.reflect.jvm.jvmName
 
 /**
- * A typealias representing a data structure used to hold extra information for multimedia activities.
+ * Representing a data structure used to hold extra information for multimedia activities.
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt:41]:
```suggestion
 * Required information for multimedia activities.
```

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt [hunk 19]
```diff
@@ -78,15 +77,10 @@ class MultimediaActivity : AnkiActivity() {
...
-//            arguments = bundleOf(
-//                MULTIMEDIA_ARGS_EXTRA to intent.extras?.getSerializableCompat<MultimediaActivityExtra>(MULTIMEDIA_ARGS_EXTRA)
-//            )
-//        }
-
         val fragment = getInstanceFromClassName<Fragment>(fragmentClassName).apply {
             arguments = bundleOf(
-                MULTIMEDIA_ARGS_EXTRA to intent.extras?.getSerializableCompat<MultimediaActivityExtra>(MULTIMEDIA_ARGS_EXTRA)
+                MULTIMEDIA_ARGS_EXTRA to intent.extras?.getSerializableCompat<MultimediaActivityExtra>(MULTIMEDIA_ARGS_EXTRA),
+                EXTRA_MEDIA_OPTIONS to intent.getSerializableExtraCompat<Serializable>(EXTRA_MEDIA_OPTIONS)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt:83]: style: consider an extension property on `Intent`, these lines are long and hard to quickly parse

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt [hunk 20]
```diff
@@ -107,14 +101,19 @@ class MultimediaActivity : AnkiActivity() {
         const val MULTIMEDIA_RESULT = "multimedia_result"
         const val MULTIMEDIA_RESULT_FIELD_INDEX = "multimedia_result_index"
 
+        // used in case a fragment supports more than media operations
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt:104]: Use `/* */` for IDE documentation

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt [hunk 21]
```diff
@@ -107,14 +101,19 @@ class MultimediaActivity : AnkiActivity() {
...
+        const val EXTRA_MEDIA_OPTIONS = "extra_media_options"
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt:105]: nitpick: architecture

Does this work from an implementation perspective? 

Typically you want to model options as a tuple\*: `(Class, Options)`, so you can't assign the wrong options to a fragment

Here, it feels like the options should be added/handled in the subclasses

`*` or similar to a tuple: `Fragment<TOptions>`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt [hunk 22]
```diff
@@ -0,0 +1,411 @@
...
+     * @receiver MediaOption The media option for which the title string is to be returned.
+     * @param context Context The context to use for accessing string resources.
+     * @return String The title string corresponding to the current media option.
+     */
+    private fun MediaOption.toTitle(context: Context): String {
+        return when (this) {
+            AUDIO_CLIP -> context.getString(R.string.multimedia_editor_popup_audio_clip)
+            VIDEO_CLIP -> context.getString(R.string.multimedia_editor_popup_video_clip)
+        }
+    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/AudioVideoFragment.kt:410]: nit: In this case, make it a method

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt [hunk 23]
```diff
@@ -349,77 +424,25 @@ class MultimediaImageFragment : MultimediaFragment(R.layout.fragment_multimedia_
...
                 MultimediaImageFragment::class,
-                multimediaExtra
+                multimediaExtra,
+                imageOptions
             )
         }
     }
 
-    /** Enum class that represents image options that a user choose from the bottom sheet  **/
+    /** Represents image options that a user choose from the bottom sheet which [MultimediaImageFragment] uses **/
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaImageFragment.kt:443]:
```suggestion
    /** Image options that a user choose from the bottom sheet which [MultimediaImageFragment] uses **/
```

**Comment by david-allison** [comment]
I consider this a "merge when you feel it's ready" @criticalAY. 

All my comments [besides the message in Discord about fixing the commit message] were nitpicks, and we can get this in now if you'd like

File: AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt (hunk 21 above)
Comment (criticalAY) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt:105]: Not clearly getting it but the getItntent in the fragment will accept only the option that are in enum it's serializable here but in fragment we have restricted the option

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimedia/MultimediaActivity.kt:111]: More of a teaching experience than anything, we can skip this

**Comment by criticalAY** [comment]
CI is happy we can get it in

**Comment by david-allison** [comment]
* https://github.com/ankidroid/Anki-Android/labels/Strings : https://github.com/ankidroid/Anki-Android/pull/16793

**Review by david-allison — APPROVED** [review]

**Comment by david-allison** [comment]
Looks like GitHub threw a wobbly. Let's get this in

**Pull request merged**